### ISSUE TITLE
getting nan as a loss when finetuning with Mono 

### ISSUE THREAD

**Issue by zCompKid** [nianticlabs/monodepth2#160]
Hello, 
Thanks for the amazing work. 
I finetuned the pre-trained model on my dataset and it worked perfectly but when I tried to train it on Waymo dataset (resized to 800x ). Apparently, the loss is _nan_ and while debugging I found out that it is in the SSIM function and it works when using --no_ssim. The values of mu_x and mu_y are sometimes either large numbers or nan. What could be the possible solution?

Thanks!

**Comment by daniyar-niantic** [comment]
Hi @zCompKid 

Could you check that your images are in 0 to 1 range?

**Comment by zCompKid** [comment]
Thanks @daniyar-niantic for the quick reply. 
No, the pred image is not always in the 0 to 1 range. It has large numbers sometimes.

**Comment by daniyar-niantic** [comment]
My question was about input images, so negative numbers in prediction images suggest that there is something wrong with the input images.

**Comment by zCompKid** [comment]
The input images are in the 0 to 1 range.  FYI, I am not using stereo for the Waymo.
I tried to replace the images with my images (only mono) but it is the same.

**Comment by zCompKid** [comment]
@daniyar-niantic I was wondering what solutions I can try to fix this? Thanks in advance!

**Comment by daniyar-niantic** [comment]
I cannot image how you could have negative values in "pred images" if your input images are in the 0 to 1 range. Could you find images that cause "nan" loss and show them here?

**Comment by zCompKid** [comment]
![000000000000396](https://user-images.githubusercontent.com/51483466/78154557-b0b91200-743c-11ea-8bfc-dbef03bb463c.jpg)
![000000000000397](https://user-images.githubusercontent.com/51483466/78154564-b1ea3f00-743c-11ea-9711-e720db60523a.jpg)
![000000000000398](https://user-images.githubusercontent.com/51483466/78154567-b1ea3f00-743c-11ea-8fba-61e0585b9549.jpg)
These are example of the dataset.

**Comment by daniyar-niantic** [comment]
Thanks for the examples. Is this the triplet of frames that causes nan loss, or is it just a random sample of the dataset?

**Comment by zCompKid** [comment]
It is the triplet of frames.

**Comment by daniyar-niantic** [comment]
Sorry to add more work, but is it possible to also share "pred images" for this (or other nan-y) triplet of frames?

**Comment by zCompKid** [comment]
Dear @daniyar-niantic, these are some of the frames and the associated pred and disp images in the first epoch. Thank you!
![waymo_fig1](https://user-images.githubusercontent.com/51483466/78275226-9a30ba80-7511-11ea-85c7-3bd6f061e9a8.png)
![waymo_fig2](https://user-images.githubusercontent.com/51483466/78275232-9ac95100-7511-11ea-8647-522442dc7d58.png)
![waymo_fig3](https://user-images.githubusercontent.com/51483466/78275235-9b61e780-7511-11ea-8a8d-6785a6dce1e6.png)
![waymo_fig4](https://user-images.githubusercontent.com/51483466/78275237-9b61e780-7511-11ea-8992-c3eb2202cd0e.png)

**Comment by mrharicot** [comment]
Hi @zCompKid , 
I have a few questions for you:

- Could you share the code for your dataloader?
The predicted images should not be able to have values outside the range 0, 1 which means the input images already have invalid values.
- Did you extract the frames yourself from the waymo data? the compression is very high which is a problem as monodepth is trained with a photometric loss and compression artefacts will cause issues.

**Comment by zCompKid** [comment]
@mrharicot Here is the code for my dataloader. I have extracted the data myself and resized the frames from (1920x1280) to (864x576).

    class WaymoDataset(MonoDataset):
        """Superclass for different types of Graz dataset loaders
        """
        def __init__(self, *args, **kwargs):
            super(WaymoDataset, self).__init__(*args, **kwargs)

            self.K = np.array([[2.070548e+03/1920, 0, 9.582694e+02, 0],
                            [0, -2.070548e+03/1280, 6.426130e+02, 0],
                            [0, 0, 1, 0],
                            [0, 0, 0, 1]], dtype=np.float32)

            self.full_res_shape = (864, 576)
            self.side_map = {"FRONT": 0, "FRONT_LEFT": 1, "FRONT_RIGHT": 2, "SIDE_LEFT": 3, "SIDE_RIGHT": 4}

        def check_depth(self):
            return False
            line = self.filenames[0].split()
            #scene_name = line[0]
            frame_index = int(line[1])

            depth_filename = os.path.join(
                self.data_path, "depth",
                "{}.npy".format(str(frame_index).zfill(INDEX_LENGTH)))

            return os.path.isfile(depth_filename)

        def check_disp(self):
            return False
            line = self.filenames[0].split()
            #scene_name = line[0]
            frame_index = int(line[1])

            depth_filename = os.path.join(
                self.data_path, "disp",
                "img_{}_disp.npy".format(int(frame_index)))

            return os.path.isfile(depth_filename)

        def get_color(self, folder, frame_index, side, do_flip):
            sides = {'l': 0, 'r':1}
            color = self.loader(self.get_image_path(folder, frame_index, sides[side]))

            if do_flip:
                color = color.transpose(pil.FLIP_LEFT_RIGHT)

            return color


    class WaymoSimpleDataset(WaymoDataset):
        """Graz dataset which loads the images from the static setup
        """
        def __init__(self, *args, **kwargs):
            super(WaymoSimpleDataset, self).__init__(*args, **kwargs)

        def get_image_path(self, folder, frame_index, side):
            # side_str = {0: "image_1", 1: "image_2"}
            f_str = "{}{}".format(str(frame_index).zfill(INDEX_LENGTH), self.img_ext)
            image_path = os.path.join(
                self.data_path, folder, "RGB", f_str)

            return image_path

        def get_depth(self, folder, frame_index, side, do_flip=False):
            self.depth_ext = '.npy'
            f_str = "{}{}".format(str(frame_index).zfill(INDEX_LENGTH), self.depth_ext)
            depth_path = os.path.join(
                self.data_path, folder, "depth", f_str)

            depth_gt = np.load(depth_path)

            if do_flip:
                depth_gt = np.fliplr(depth_gt)

            return depth_gt

        def get_disparity(self, folder, frame_index, side, do_flip=False):
            self.disp_ext = '.npy'
            f_str = "img_{}_disp{}".format(frame_index, self.disp_ext)
            disp_path = os.path.join(
                self.data_path, folder, "disp", f_str)

            disp_gt = np.load(disp_path)

            if do_flip:
                disp_gt = np.fliplr(disp_gt)

            return disp_gt

**Comment by mrharicot** [comment]
Bingo! Your intrinsics matrix K is incorrect.
You forgot to normalize the principal point location, and the vertical focal length is somehow negative?
Here's the corrected one:
```python
self.K = np.array([[2.070548e+03/1920, 0, 9.582694e+02/1920, 0],
                        [0, 2.070548e+03/1280, 6.426130e+02/1280, 0],
                        [0, 0, 1, 0],
                        [0, 0, 0, 1]], dtype=np.float32)
```
Furthermore I would recommend reexporting the data and save the frames with less compression (jpeg 95% is a good place to start). You might also want to export them at a lower resolution to speed up training significantly, 864x576 is very high. I would recommend to start with something smaller like 480x320.

I hope this helps!

**Comment by zCompKid** [comment]
@mrharicot, @daniyar-niantic  Thank you so much! 
My bad.. The pred image looks good now. I will let you know about the result after the training.

**Comment by mrharicot** [comment]
Great to hear! Don't

**Comment by mrharicot** [comment]
Closing now that the main issue has been solved.
@zCompKid let us know how it goes. But I would really strongly recommend that your reexport your data, it will take a very long time to train, and the quality of your depth won't be great given how compressed your images are.

**Comment by zCompKid** [comment]
@mrharicot The pred image is fixed but the loss is still _NaN_ :(

**Comment by daniyar-niantic** [comment]
Could you find the triplet causing the (new) nan issue and the pred image again?

**Comment by zCompKid** [comment]
Here are some of the examples that cause NaN loss. However, it is not associated with the images: it is different images with different runs.  
The values of the pred images range from a large negative number to a large positive number.

![waymo_fig10](https://user-images.githubusercontent.com/51483466/78301977-3f5f8900-753a-11ea-9270-a0fa0b5ddd74.png)
![waymo_fig11](https://user-images.githubusercontent.com/51483466/78301979-3ff81f80-753a-11ea-876c-2bf149620517.png)

**Comment by daniyar-niantic** [comment]
Your geometry is still broken. Your pred image has pixels from sky above and below the center of the image. Maybe you still have problems with your intrinsics?

**Comment by zCompKid** [comment]
I got the intrinsics from the Waymo data.

**Comment by mrharicot** [comment]
Hmm something's still not right.
What does your `train_files.txt` look like?

**Comment by zCompKid** [comment]
here is the sample: 

     left 2 l
     left 3 l
     left 4 l
     left 5 l
     ...

**Comment by mrharicot** [comment]
Here are some more questions:
- How many frames total are you using? 
- From how many sequences?
- At what framerate?

**Comment by zCompKid** [comment]
I am using 1987 frames from 10 sequences. I am not used about the framerate but I think each sequence is 20 or 30 seconds long.

**Comment by zCompKid** [comment]
I think the problem was that I was trying to finetune the pose network of the pre-trained model. It is working now after fine-tuning only the encoder and depth decoder and train the pose net from scratch. 
Thank you so much for the helpful support!

**Comment by mrharicot** [comment]
Glad to hear you got it working!
